src/mlem.rs

```diff
@@ -216,6 +220,25 @@ impl Image {
     }
 }
 
+/// Define the scatter prediction histogram (incomplete)
+pub fn scatter_prediction<T>(lors: &[LOR]) -> Option<Scattergram<T>>
+where
+    T: Lorogram
+{
+    // This function is not really what I want but it gets me started.
+    // Should probably allow type of Lorogram as parameter. JustZ for now
+    let scatters = Scattergram::new(JustZ::new(2000.0, 20));
+
+    // Just call every 10th LOR scatter for now.
+    for (n, lor) in lors.iter().enumerate() {
+        let lor_type = if n % 10 == 0 {Prompt::Scatter} else {Prompt::True};
+        scatters.fill(lor_type, lor.p1, lor.p2);
```

Comment:
Even though the `p1` and `p2` variables of `LOR` are type `Point` and `fill` expects `Point` this gives an error. A bit lost here since I can't find the implementation of `Point` to even extract the values to a tuple which seems to be acceptable even though the function definition says otherwise.

Reply:
This function has `LOR` in its signature. LSP takes me to the definition of `LOR` (in Emacs it's `M-.` on top of `LOR`, no idea what it is in VSCode, but it *should* work somehow). Once there, I see `Point` as the type of the `p1` and `p2` members. So I ask LSP to take me to the definition of `Point`, and it brings me [here](https://github.com/andLaing/petalo/blob/apply-lorogram/src/types.rs#L17). That's one side of the puzzle.

The other side is the `Point` type accepted by `fill`. Going to the declaration of `fill` in `Lorogram` or the any of its `impl`s and asking LSP to take me to the definition of `Point` takes me [here](https://github.com/andLaing/petalo/blob/apply-lorogram/src/lorogram/mod.rs#L5).

Additionally, hovering over either of these `Point`s should show you what they are, and give you enough information. 

If any of these things do not work for you in VSCode, then there is something seriously wrong, and it needs fixing.

Reply:
Given that `Lorogram` is a **LOR**ogram, it's reasonable that its methods should accept a `LOR`s rather than points as  parameters. I didn't do that in the prototype, because I didn't want to couple the code tightly to the `LOR` type, especially when experimenting. But in the end the `Lorogram` functions will probably take `LOR`s, so this particular difficulty will just go away.

Reply:
I had followed the first but hadn't noticed that there was a separate `Point` defined, a bit unstable, surely?

In any case, the first Point doesn't tell me how to get the values of xyz and googling the `nc::math ::Point` type doesn't help much. I've found some examples where the values are accessed I'll see what I can do with that.

Reply:
> a bit unstable, surely?

Two things:

Firstly, be aware that you are targeting completely experimental, exploratory, hacky code, so you must expect to find bad,  random, not well-thought-out choices all over the place, which can be changed without notice. (This is one strong argument in favour of doing this con peras, rather than on your own.)

Secondly, the namespace rules are very 'stable' and very clear. It's perfectly OK to have two things with the same name in different namespaces (that's why namespaces exist), and you just have to get into the habit of being aware that there's more than just the 'leaf' of the name. This comes with both advantages and disadvantages, but that's life when writing complex software.

> googling `the nc::math ::Point` type doesn't help much.

It wouldn't, because `nc` is the local name for `ncollide3d`. That's why it's important to get into the habit of getting LSP to give you such crucial details, rather than just looking at the superficial information.

> In any case, the first Point doesn't tell me how to get the values of xyz

Yes, in this case the experience is rather disappointing. Sometimes the standard Rust docs format with its search box is very useful, sometimes it's very frustrating. This appears to be an instance of the latter.

Don't forget the wonders of LSP, though: frequently you just stick a dot after the object and type a couple of characters you guess might be relevant, and the thing you're looking for magically shows up. In this case it doesn't happen, because  this `Point` has a million methods which contain `x`, `y` and `z` in their names, so the thing you're looking for (public members called `x`, `y` and `z`) are lost in the noise (actually, maybe they don't show up at all?).

Reply:
> It wouldn't, because nc is the local name for ncollide3d. That's why it's important to get into the habit of getting LSP to give you such crucial details, rather than just looking at the superficial information.

Yeah, I meant to imply the `ncollide3d` and I googled after following the definitions down the rabbit hole. Frustrating.

I've pushed a very hacky overly verbose version, without flexibility which compiles (but probably doesn't do much of use). I need to understand traits a lot more before I'll be able to implement the use of `Scattergram` of "any" `Lorogram` but never mind, first day of RUST coding isn't a complete loss.

Reply:
Related RUST question:
I tried `let Point{x: x1, y: y1, z: z1, ..} = lor.p1` which failed because:
`struct `OPoint` does not have fields named `x`, `y`, `z``
What am I missing? I thought it'd work since `lor.p1.x` works, for example.

Reply:
Dunno. There's some magic going on there that I don't understand. If you follow the trail, you end up at the `Matrix` type which seems to be the foundation of all this stuff:

https://docs.rs/nalgebra/latest/src/nalgebra/base/matrix.rs.html#162-199

It states that you can access components using field notation, but I can't find the actual fields anywhere (which might explain why LSP doesn't propose them ... and why can't pattern match them). This leads me to suspect that there might be some magic going on similar to Python's properties. If something like that exists in Rust, then I've never come across it.

The crates in this family (https://www.dimforge.com/) seem very useful and powerful, but they do regularly leave me scratching my head.

Given how little of their power we actually use, I keep being tempted to get rid of them and just implement the functionality myself. The fact that they don't work with `uom` is further motivation. On the other hand, I do want to accelerate the code with SIMD (one day) and these crates are SIMD-aware, so it would be good to get that for free. I wonder what the cost of 'free' is, in this case.

Reply:
Yes, it turns out that these crates are too clever by half. `Point` is a type synonym for `OPoint` and the latter has some magic which can *implicitly* convert it to things like [`X`](https://docs.rs/nalgebra/latest/nalgebra/base/coordinates/struct.X.html), which provide access via fields like `.x`.

Very clever and very confusing.